### DISCUSSION TITLE
Should a tooltip show on a focus event generated programmatically?

### DISCUSSION THREAD

**Discussion by gnapse** [ariakit/ariakit#749] [Questions] [answered]
The reason for this question is better expressed via a short screen recording:

![CleanShot 2020-09-23 at 23 15 07](https://user-images.githubusercontent.com/15199/94093060-adeb1880-fdf2-11ea-9dd6-0a2f3cd2e1d9.gif)

I open a menu. However, when I click outside the menu the focus is automatically set back to the `MenuButton`. This then triggers the tooltip to appear, as it is usually expected when a button with a tooltip receives focus.

However, I'd argue that in this case the tooltip appearance is a nuisance. This means that all menu buttons with a tooltip will have their tooltip appear when the menu is dismissed. Even clicking a menu option will leave a tooltip floating around. This is presumably true for all buttons that trigger a popover or even a dialog.

I first wanted to know if this is a reasonable expectation on my side, and if so, then I'd like to discuss solutions. I was hoping this is something I could fix in user-land without the library having to fix it for me.

**Comment by diegohaz** [comment, marked as answer]
> I first wanted to know if this is a reasonable expectation on my side, and if so, then I'd like to discuss solutions. I was hoping this is something I could fix in user-land without the library having to fix it for me.

I think that is a really reasonable expectation.

As for the solution:

I think both this and tooltip delays (that we're experimenting with the `unstable_timeout` option) are part of the same feature: intent. We should set heuristics to determine whether the user has intention to see the tooltip.

With mouse over, delaying the tooltip for some milliseconds seems to do a pretty good job.

But I still don't know how to approach it for keyboard focus. We can't simply use "programmatic focus" as the criteria. For example, moving focus between items in a toolbar is technically a programmatic focus, but we do want it to trigger the tooltip.

We could maybe pass some metada to the element together with `element.focus()` when we don't want to show the tooltip. For example, when the dialog closes:
https://github.com/reakit/reakit/blob/c4c57cf43f22c8769dd439b0c989af4af79e1a83/packages/reakit/src/Dialog/__utils/useFocusOnHide.ts#L43

But, ideally, the solution should be scoped to the `Tooltip` component so it would work with dialogs from other libraries as well. I haven't tested it, but I guess something like this may work in userland already:
```jsx
<TooltipReference
  onFocus={(event) => {
    // If focus is not followed by a key up event, does it mean that it's not
    // an intentional keyboard focus? Not sure.
    const handleKeyUp = (e) => e.key !== "Escape" && tooltip.show();
    event.currentTarget.addEventListener("keyup", handleKeyUp, { once: true });
    // Prevent tooltip.show from being called by TooltipReference
    event.preventDefault();
  }}
/>
```

**Reply by gnapse** [reply]
Thanks for the thorough reply. I'll test that and let you know.

**Reply by gnapse** [reply]
It seems to work like a charm. Thanks!

**Comment by gnapse** [comment]
Actually, there's still a way in which it can appear when not wanted: by activating a menu option via the keyboard. But on the other hand, pressing Escape to close the menu works as expected.

Event with this limitation I'll take the proposed solution. My main concern was that mouse centric users would complain. But this really does fix it when interacting solely with the mouse. I may explore fixing it fully but for now it looks good. Thanks again!

**Reply by diegohaz** [reply]
Nice! I'll create an issue for this conversation so we can look into implementing this functionality directly in the library in a more consistent way.
